glib/src/variant.rs

```diff
@@ -471,22 +493,49 @@ impl Variant {
     #[doc(alias = "g_variant_new_maybe")]
     pub fn from_maybe<T: StaticVariantType>(child: Option<&Variant>) -> Self {
         let type_ = T::static_variant_type();
-        let ptr = match child {
+        match child {
             Some(child) => {
                 assert_eq!(type_, child.type_());
 
-                child.to_glib_none().0
+                Self::from_some(child)
             }
-            None => std::ptr::null(),
-        };
+            None => Self::from_none(&type_),
+        }
+    }
+
+    // rustdoc-stripper-ignore-next
+    /// Creates a new maybe Variant from a child.
+    #[doc(alias = "g_variant_new_maybe")]
+    pub fn from_some(child: &Variant) -> Self {
         unsafe {
             from_glib_none(ffi::g_variant_new_maybe(
-                type_.as_ptr() as *const _,
-                ptr as *mut ffi::GVariant,
+                ptr::null(),
+                child.to_glib_none().0,
             ))
         }
     }
 
+    // rustdoc-stripper-ignore-next
+    /// Creates a new maybe Variant with Nothing.
+    #[doc(alias = "g_variant_new_maybe")]
+    pub fn from_none(type_: &VariantTy) -> Self {
+        unsafe {
+            from_glib_none(ffi::g_variant_new_maybe(
+                type_.to_glib_none().0,
+                ptr::null_mut(),
+            ))
+        }
+    }
+
+    // rustdoc-stripper-ignore-next
+    /// Extract the value of a maybe Variant.
+    #[inline]
+    pub fn as_maybe(&self) -> Option<Variant> {
+        assert!(self.type_().is_maybe());
+
+        self.try_child_value(0)
```

Comment:
I may have misunderstood the logic, but I'm not exactly sure that this matches what the docstring says.
Let's maybe cover this method with a test to avoid any doubts?

Reply:
A maybe variant is a container with either 1 or 0 values, thus roughly the logic for this function is `if self.n_children() == 0 { None } else { Some(self.child_value(0)) }` - `Variant::try_child_value` already exists as a rough equivalent to rust's `slice.get(0)` for this purpose, so it didn't seem worth duplicating the logic here.
However, I really should just use `ffi::g_variant_get_maybe` instead, since it's clearer - [it implements the exact same logic regardless](https://gitlab.gnome.org/GNOME/glib/-/blob/main/glib/gvariant.c#L698-703).

Good point on adding a test for this method though, will do!